File: task/graph/dfs/DiameterOfBinaryTree.java
package graph.dfs;

import graph.TreeNode;

public class DiameterOfBinaryTree {
    public static void main(String[] args) {
        TreeNode treeNode = new TreeNode(1, new TreeNode(2, new TreeNode(4), new TreeNode(5)), new TreeNode(3));
        TreeNode treeNode1 = new TreeNode(1, new TreeNode(2), null);

        System.out.println(diameterOfBinaryTree(treeNode));
    }

    static int diameter = 0;

    public static int diameterOfBinaryTree(TreeNode root) {
        diameterSearchRec(root);
        return diameter;
    }

    public static int diameterSearchRec(TreeNode root) {
        if (root == null) return -1;

        int diamL = diameterSearchRec(root.left);
        int diamR = diameterSearchRec(root.right);

        diameter = Math.max(diameter, diamL + diamR + 2);

        return Math.max(diamL, diamR) + 1;
    }
}

File: python/week2/DiameterOfBinaryTree.py
from typing import Optional

from python.TreeNode import TreeNode


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def dfs(root: Optional[TreeNode]):
            nonlocal max_diameter

            if not root:
                return 0

            left_height = dfs(root.left)
            right_height = dfs(root.right)

            max_diameter = max(max_diameter, left_height + right_height)

            return 1 + max(left_height, right_height)

        max_diameter = 0
        dfs(root)
        return max_diameter


if __name__ == '__main__':
    res1 = Solution().diameterOfBinaryTree(TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3)))
    print(res1)
